File: bestia-client/src/Bnet/Message/Shop/ShopTradeCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Shop
{
  /// <summary>
  /// Buys or sells one item with the settlement we are standing in.
  /// </summary>
  /// <remarks>
  /// No price on the wire in either direction. We say how much of what, and the server quotes - so a
  /// window that has gone stale pays the current price rather than the one it is showing.
  /// </remarks>
  public partial class ShopTradeCMSG : ICMSG
  {
    public ulong ItemId { get; set; }

    public int Amount { get; set; }

    /// <summary>True when we are handing goods over, false when taking them.</summary>
    public bool Selling { get; set; }

    public ulong MerchantEntityId { get; set; }

    public override Envelope ToEnvelope()
    {
      return new Envelope
      {
        ShopTrade = new global::Bnet.ShopTradeCMSG
        {
          ItemId = ItemId,
          Amount = Amount,
          Selling = Selling,
          MerchantEntityId = MerchantEntityId
        }
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/economy/shop/ShopTradeCMSG.kt
package net.bestia.zone.economy.shop

import net.bestia.bnet.proto.ShopTradeCmsgProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.util.EntityId

/**
 * A request to buy or sell with the settlement the sender is standing in.
 *
 * No price on the wire in either direction. A client that could name its own price would be a client
 * that could be wrong about it - it says how much of what, and the server quotes.
 */
data class ShopTradeCMSG(
  override val playerId: Long,
  val itemId: Long,
  val amount: Int,
  val selling: Boolean,
  val merchantEntityId: EntityId,
) : CMSG {

  companion object {
    fun fromBnet(playerId: Long, bnet: ShopTradeCmsgProto.ShopTradeCMSG): ShopTradeCMSG {
      return ShopTradeCMSG(playerId, bnet.itemId, bnet.amount, bnet.selling, bnet.merchantEntityId)
    }
  }
}
